### ISSUE TITLE
"Error Cleaning Up Jobs" From Cronjob, Which is Not Really an Error

### ISSUE THREAD

**Issue by craig-riecke** [kubernetes/kubernetes#103607]
#### What happened:

This started happening when we upgraded to Kubernetes 1.21. The following message occurs:

```
Jul 08 07:00:01        1 cronjob_controllerv2.go:193] "error cleaning up jobs" cronjob="default/vet-messenger-batch-prod-hourly" resourceVersion="160274892" err="Operation cannot be fulfilled on cronjobs.batch \"vet-messenger-batch-prod-hourly\": the object has been modified; please apply your changes to the latest version and try again"
Jul 08 07:00:01        1 cronjob_controllerv2.go:154] error syncing CronJobController default/vet-messenger-batch-prod-hourly, requeuing: Operation cannot be fulfilled on cronjobs.batch "vet-messenger-batch-prod-hourly": the object has been modified; please apply your changes to the latest version and try again
```

The Cronjob is pretty simple:

```
apiVersion: batch/v1
kind: CronJob
metadata:
  name: vet-messenger-batch-prod-hourly
spec:
  schedule: "0 * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
            - name: vet-messenger-batch-prod
              image: REDACTED/vet-messenger:prod
              command: ["yarn", "batch:start:prod"]
              imagePullPolicy: Always
              env:
                - name: AWS_ACCESS_KEY_ID
                  valueFrom:
                    secretKeyRef:
                      name: k8s-container-runner-prod
                      key: aws_access_key_id
                - name: AWS_SECRET_ACCESS_KEY
                  valueFrom:
                    secretKeyRef:
                      name: k8s-container-runner-prod
                      key: aws_secret_access_key
                - name: AWS_REGION
                  value: us-east-1
          restartPolicy: Never
      backoffLimit: 1
```

#### What you expected to happen:

No error.  But in fact, it does appear jobs ARE being removed properly regardless of the error message.  

#### How to reproduce it (as minimally and precisely as possible):

Use the above.

#### Anything else we need to know?:

We tried deleting and recreating the cronjob both with `kubectl create -f` and `kubectl apply -f` and neither seems to work. Note the API version was at batch/v1beta1 before we upgraded, and we did change it to batch/v1 to squelch the deprecation notice.

#### Environment:
- Kubernetes version (use `kubectl version`): 
```Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:52:14Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:53:14Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`): From nodes (master is identical):
```
NAME="Ubuntu"
VERSION="20.04.2 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```
- Kernel (e.g. `uname -a`): `Linux ip-10-92-217-104 5.4.0-1047-aws #49-Ubuntu SMP Wed Apr 28 22:47:04 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: Kops
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by kerthcet** [comment]
i think it's due to the upgrade of your cluster. The error means resource version can not match? what's your cluster version before?  @craig-riecke 

your new created cronjob is working well, right? like below:
```
apiVersion: batch/v1
kind: CronJob
metadata:
  name: hello
spec:
  schedule: "*/1 * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: hello
            image: busybox
            imagePullPolicy: IfNotPresent
            command:
            - /bin/sh
            - -c
            - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure
```

**Comment by craig-riecke** [comment]
The previous cluster version was 1.18, and yes I know we did a bad thing by skipping two minor releases.  

The newly created cronjob is working fine. And strangely, there are no old jobs that `vet-messenger-batch-prod-hourly` spawned - except for the three most recently run, all of them obviously created in the new cluster (we use the defaults here).  So jobs ARE getting cleaned up.  That's what makes the error messages baffling.

**Comment by kerthcet** [comment]
`successfulJobsHistoryLimit` is set to default 3, `.spec.failedJobsHistoryLimit` is set to default 1.
see https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/#jobs-history-limits for details.

**Comment by kerthcet** [comment]
`successfulJobsHistoryLimit` is set to default 3, `.spec.failedJobsHistoryLimit` is set to default 1.

**Comment by caleb15** [comment]
We are also seeing this error message in our logs.

Some of our cronjobs say that there's an active job running, yet the job referenced has already been completed. Could that be a result of this bug?

For example:

> kubectl describe job check-saml2-metadata-27222245
```
Name:           check-saml2-metadata-27222245
Namespace:      cloud20
Selector:       controller-uid=864e4340-c9a8-45cb-a035-b9dac399cbf6
Labels:         instance=check-saml2-metadata
                name=cron
                tags.datadoghq.com/env=cloud20
                tags.datadoghq.com/service=cron
                tags.datadoghq.com/version=432b9e0
Annotations:    <none>
Controlled By:  CronJob/check-saml2-metadata
Parallelism:    1
Completions:    1
Start Time:     Mon, 04 Oct 2021 01:05:00 -0700
Completed At:   Mon, 04 Oct 2021 01:05:43 -0700
Duration:       43s
Pods Statuses:  0 Running / 1 Succeeded / 0 Failed
```

As you can see it successfully completed. Yet the cronjob lists it as active:
> kubectl describe cronjob check-saml2-metadata
```
Active Jobs:         check-saml2-metadata-27222245
```

**Comment by pere3** [comment]
> We are also seeing this error message in our cloudwatch logs.
> 
> Some of our cronjobs say that there's an active job running, yet the job referenced has already been completed. Could that be a result of this bug?
> 
> For example:
> 
> > kubectl describe job check-saml2-metadata-27222245
> 
> ```
> Name:           check-saml2-metadata-27222245
> Namespace:      cloud20
> Selector:       controller-uid=864e4340-c9a8-45cb-a035-b9dac399cbf6
> Labels:         instance=check-saml2-metadata
>                 name=cron
>                 tags.datadoghq.com/env=cloud20
>                 tags.datadoghq.com/service=cron
>                 tags.datadoghq.com/version=432b9e0
> Annotations:    <none>
> Controlled By:  CronJob/check-saml2-metadata
> Parallelism:    1
> Completions:    1
> Start Time:     Mon, 04 Oct 2021 01:05:00 -0700
> Completed At:   Mon, 04 Oct 2021 01:05:43 -0700
> Duration:       43s
> Pods Statuses:  0 Running / 1 Succeeded / 0 Failed
> ```
> 
> As you can see it successfully completed. Yet the cronjob lists it as active:
> 
> > kubectl describe cronjob check-saml2-metadata
> 
> ```
> Active Jobs:         check-saml2-metadata-27222245
> Events:
>   Type    Reason            Age                  From                Message
>   ----    ------            ----                 ----                -------
>   Normal  JobAlreadyActive  12m (x24 over 3d5h)  cronjob-controller  Not starting job because prior execution is running and concurrency policy is Forbid
> ```
> 
> Relevant Cloudwatch logs:
> 
> ```
> I1004 08:05:00.316858 1 event.go:291] "Event occurred" object="cloud20/check-saml2-metadata-27222245" kind="Job" apiVersion="batch/v1" type="Normal" reason="SuccessfulCreate" message="Created pod: check-saml2-metadata-27222245-kw2h7"
> I1004 08:05:00.362801 1 cronjob_controllerv2.go:193] "error cleaning up jobs" cronjob="cloud20/check-saml2-metadata" resourceVersion="5137310" err="Operation cannot be fulfilled on cronjobs.batch \"check-saml2-metadata\": the object has been modified; please apply your changes to the latest version and try again"
> E1004 08:05:00.362816 1 cronjob_controllerv2.go:154] error syncing CronJobController cloud20/check-saml2-metadata, requeuing: Operation cannot be fulfilled on cronjobs.batch "check-saml2-metadata": the object has been modified; please apply your changes to the latest version and try again
> I1004 08:05:43.264713 1 event.go:291] "Event occurred" object="cloud20/check-saml2-metadata-27222245" kind="Job" apiVersion="batch/v1" type="Normal" reason="Completed" message="Job completed"
> ```

seeing the same thing on GKE 1.21.5-gke.1302

**Comment by pere3** [comment]
@RyanAoh hi! any news on that PR? it seems there is a need for rebase

**Comment by pere3** [comment]
The whole thing with JobAlreadyActive is that it prevents next schedules of a Job, since we have `Concurrency Policy:            Forbid` as our main mechanism to prevent overlapping.

**Comment by RyanAoh** [comment]
@pere3  Thanks for reminding me. I've updated the PR.

**Comment by pere3** [comment]
> @pere3 Thanks for reminding me. I've updated the PR.

can you please confirm, that this bug is what causing CronJobs to stuck on JobAlreadyActive reason?

**Comment by RyanAoh** [comment]
> We are also seeing this error message in our cloudwatch logs.
> 
> Some of our cronjobs say that there's an active job running, yet the job referenced has already been completed. Could that be a result of this bug?
> 
> For example:
> 
> > kubectl describe job check-saml2-metadata-27222245
> 
> ```
> Name:           check-saml2-metadata-27222245
> Namespace:      cloud20
> Selector:       controller-uid=864e4340-c9a8-45cb-a035-b9dac399cbf6
> Labels:         instance=check-saml2-metadata
>                 name=cron
>                 tags.datadoghq.com/env=cloud20
>                 tags.datadoghq.com/service=cron
>                 tags.datadoghq.com/version=432b9e0
> Annotations:    <none>
> Controlled By:  CronJob/check-saml2-metadata
> Parallelism:    1
> Completions:    1
> Start Time:     Mon, 04 Oct 2021 01:05:00 -0700
> Completed At:   Mon, 04 Oct 2021 01:05:43 -0700
> Duration:       43s
> Pods Statuses:  0 Running / 1 Succeeded / 0 Failed
> ```
> 
> As you can see it successfully completed. Yet the cronjob lists it as active:
> 
> > kubectl describe cronjob check-saml2-metadata
> 
> ```
> Active Jobs:         check-saml2-metadata-27222245
> Events:
>   Type    Reason            Age                  From                Message
>   ----    ------            ----                 ----                -------
>   Normal  JobAlreadyActive  12m (x24 over 3d5h)  cronjob-controller  Not starting job because prior execution is running and concurrency policy is Forbid
> ```
> 
> Relevant Cloudwatch logs:
> 
> ```
> I1004 08:05:00.316858 1 event.go:291] "Event occurred" object="cloud20/check-saml2-metadata-27222245" kind="Job" apiVersion="batch/v1" type="Normal" reason="SuccessfulCreate" message="Created pod: check-saml2-metadata-27222245-kw2h7"
> I1004 08:05:00.362801 1 cronjob_controllerv2.go:193] "error cleaning up jobs" cronjob="cloud20/check-saml2-metadata" resourceVersion="5137310" err="Operation cannot be fulfilled on cronjobs.batch \"check-saml2-metadata\": the object has been modified; please apply your changes to the latest version and try again"
> E1004 08:05:00.362816 1 cronjob_controllerv2.go:154] error syncing CronJobController cloud20/check-saml2-metadata, requeuing: Operation cannot be fulfilled on cronjobs.batch "check-saml2-metadata": the object has been modified; please apply your changes to the latest version and try again
> I1004 08:05:43.264713 1 event.go:291] "Event occurred" object="cloud20/check-saml2-metadata-27222245" kind="Job" apiVersion="batch/v1" type="Normal" reason="Completed" message="Job completed"
> ```

I‘m not sure whether it is also a result of this bug since I didn't reproduce that as you described. But according to the code and logs, they seem to be related. Maybe you can retest it when this fixed.

**Comment by RyanAoh** [comment]
> > @pere3 Thanks for reminding me. I've updated the PR.
> 
> can you please confirm, that this bug is what causing CronJobs to stuck on JobAlreadyActive reason?

I'll try to take a closer look at it when I have time.

**Comment by RyanAoh** [comment]
> > @pere3 Thanks for reminding me. I've updated the PR.
> 
> can you please confirm, that this bug is what causing CronJobs to stuck on JobAlreadyActive reason?

OK, I'll try to take a closer look at it recently.

**Comment by oradwell** [comment]
@pere3 @caleb15 we have the same thing with Forbid

we use helm templates to deploy using Spinnaker so everytime we recreate cronjobs they have the old `apiVersion: batch/v1beta1`. I think they automatically get modified when cronjob objects are created to upgrade to `batch/v1`.

I didn't know that objects were automatically modified when they have the old apiVersion. I noticed Ingress objects were also modified. so I'm updating our helm charts to fix this

**Comment by ahg-g** [comment]
/cc @soltysh 

this is related to cronjob v2 controller

**Comment by oradwell** [comment]
I wrote a bash script to clean up the jobs affected by this bug
https://gist.github.com/oradwell/8d3262326dd2b7d51c63708a698e861d

- it looks for active cronjobs
- then their active jobs
- checks the actual status of the jobs
- deletes the jobs that are complete

**Comment by ahg-g** [comment]
@oradwell the fix is merged and backported to 1.21; are you able to use the latest version to test the fix? It is available in 1.21.9, 1.22.6 and 1.23.3

**Comment by oradwell** [comment]
@ahg-g thanks for confirming the patch versions, I wasn't sure what patch version it would come in. we're on the latest version of EKS (1.21.5). we'll need to wait for it to come to EKS
